File: Longest Increasing Subsequence[12-August-2023]/Binary Search Based Approach/Solution.java
// >> JAVA CODE <<

class Solution 
{
    //Function to find length of longest increasing subsequence.
    static int lowerbound(int[] arr,int size,int x)
    {
        int low=0,high=size-1;
        int ans=-1;
        while(low<=high)
        {
            int mid=(low+high)/2;
            
            if(arr[mid]>=x){
                ans=mid;
                high=mid-1;
            }
            else
                low=mid+1;
        }
        
            return ans;
    }
    static int longestSubsequence(int size, int a[])
    {
        int[] ans=new int[size];
        int sz=0;
        ans[sz++]=a[0];
        
        for(int i=1;i<size;i++)
        {
            if(ans[sz-1]<a[i])
            {
                ans[sz++]=a[i];
            }
            else
            {
                int index=lowerbound(ans,sz,a[i]);
                
                ans[index]=a[i];
            }
        }
        
            return sz;
    }
}

File: Longest Increasing Subsequence[12-August-2023]/Binary Search Based Approach/Solution.cpp
//  >> CPP CODE <<

class Solution
{
    public:
    //Function to find length of longest increasing subsequence.
    int lowerbound(vector<int> arr,int x)
    {
        int low=0,high=arr.size()-1;
        
        int ans=-1;
        
        while(low<=high)
        {
            int mid=(low+high)/2;
            
            if(arr[mid]>=x)
            {
                ans=mid;
                high=mid-1;
            }
            else
            {
                low=mid+1;
            }
        }
        
            return ans;
    }
    int longestSubsequence(int n, int a[])
    {
       vector<int> ans;
       
       ans.push_back(a[0]);
       
       for(int i=1;i<n;i++)
       {
           if(a[i]>ans.back())
           {
               ans.push_back(a[i]);
           }
           else
           {
               int low = lowerbound(ans,a[i]);
                ans[low]=a[i];
           }
           
           for(int x=0;x<ans.size();x++) cout<<ans[x]<<" ";
           cout<<endl;
           
       }
       
        
            return ans.size();
    }
};

